runtime/lua/vim/lsp.lua

```diff
@@ -1509,24 +1509,28 @@ function lsp.start_client(config)
         { method = method, client_id = client_id, bufnr = bufnr, params = params }
       )
     end, function(request_id)
-      local request = client.requests[request_id]
-      request.type = 'complete'
-      nvim_exec_autocmds('LspRequest', {
-        buffer = bufnr,
-        modeline = false,
-        data = { client_id = client_id, request_id = request_id, request = request },
-      })
+      if vim.api.nvim_buf_is_valid(bufnr) then
+        local request = client.requests[request_id]
+        request.type = 'complete'
+        nvim_exec_autocmds('LspRequest', {
+          buffer = bufnr,
+          modeline = false,
+          data = { client_id = client_id, request_id = request_id, request = request },
+        })
+      end
       client.requests[request_id] = nil
     end)
 
     if success and request_id then
-      local request = { type = 'pending', bufnr = bufnr, method = method }
-      client.requests[request_id] = request
-      nvim_exec_autocmds('LspRequest', {
-        buffer = bufnr,
-        modeline = false,
-        data = { client_id = client_id, request_id = request_id, request = request },
-      })
+      if vim.api.nvim_buf_is_valid(bufnr) then
```

Comment:
Probably not necessary, because it doesn't make much sense to call `client.request` with an invalid buffer in the first place.

Reply:
Agreed, this buffer validity check should only be in the response callback handler (one spot). The other two happen synchronously based on client-side initiated events.